format/FlightSql.proto

```diff
@@ -0,0 +1,454 @@
+ *
+ * The returned schema will be:
+ * <
+ *  pk_catalog_name: utf8,
+ *  pk_schema_name: utf8,
+ *  pk_table_name: utf8,
+ *  pk_column_name: utf8,
+ *  fk_catalog_name: utf8,
+ *  fk_schema_name: utf8,
+ *  fk_table_name: utf8,
+ *  fk_column_name: utf8,
+ *  key_sequence: int,
+ *  fk_key_name: utf8,
+ *  pk_key_name: utf8,
+ *  update_rule: uint1,
+ *  delete_rule: uint1
+ * >
+ * The returned data should be ordered by fk_catalog_name, fk_schema_name, fk_table_name, fk_key_name, then key_sequence.
+ * update_rule and delete_rule returns a byte that is equivalent to actions:
+ *    - 0 = CASCADE
+ *    - 1 = RESTRICT
+ *    - 2 = SET NULL
+ *    - 3 = NO ACTION
+ *    - 4 = SET DEFAULT
+ */
+message CommandGetExportedKeys {
+  option (experimental) = true;
+
+  // Specifies the catalog to search for the foreign key table.
+  google.protobuf.StringValue catalog = 1;
+
+  // Specifies the schema to search for the foreign key table.
+  google.protobuf.StringValue schema = 2;
+
+  // Specifies the foreign key table to get the foreign keys for.
+  string table = 3;
+}
+
+/*
+ * Represents a request to retrieve the foreign keys of a table on a Flight SQL enabled backend.
+ * Used in the command member of FlightDescriptor for the following RPC calls:
+ *  - GetSchema: return the schema of the query.
+ *  - GetFlightInfo: execute the catalog metadata request.
+ *
+ * The returned schema will be:
+ * <
+ *  pk_catalog_name: utf8,
+ *  pk_schema_name: utf8,
+ *  pk_table_name: utf8,
+ *  pk_column_name: utf8,
+ *  fk_catalog_name: utf8,
+ *  fk_schema_name: utf8,
+ *  fk_table_name: utf8,
+ *  fk_column_name: utf8,
+ *  key_sequence: int,
+ *  fk_key_name: utf8,
+ *  pk_key_name: utf8,
+ *  update_rule: uint1,
+ *  delete_rule: uint1
+ * >
+ * The returned data should be ordered by pk_catalog_name, pk_schema_name, pk_table_name, pk_key_name, then key_sequence.
+ * update_rule and delete_rule returns a byte that is equivalent to actions:
+ *    - 0 = CASCADE
+ *    - 1 = RESTRICT
+ *    - 2 = SET NULL
+ *    - 3 = NO ACTION
+ *    - 4 = SET DEFAULT
+ */
+message CommandGetImportedKeys {
+  option (experimental) = true;
+
+  // Specifies the catalog to search for the primary key table.
+  google.protobuf.StringValue catalog = 1;
+
+  // Specifies the schema to search for the primary key table.
+  google.protobuf.StringValue schema = 2;
+
+  // Specifies the primary key table to get the foreign keys for.
+  string table = 3;
+}
+
+// SQL Execution Action Messages
+
+/*
+ * Request message for the "CreatePreparedStatement" action on a Flight SQL enabled backend.
+ */
+message ActionCreatePreparedStatementRequest {
+  option (experimental) = true;
+
+  // The valid SQL string to create a prepared statement for.
+  // The query should be treated as an opaque value, that is, clients should not attempt to parse this.
+  string query = 1;
+}
+
+/*
+ * Wrap the result of a "GetPreparedStatement" action.
```

Comment:
For resource like this did we ever clarify requirements around close?  Or are the semantics left up to the server?  We should probably document one way or another.

Reply:
The prepared statement is only closed when the client calls the ClosePreparedStatement action.
Do you think it should be mentioned in this action?

Reply:
Forgot to mention that the prepared statement should be automatically closed by the server if not used after some time.

Reply:
should or may?

Reply:
Should - given that the close is a client-side initiated action, the server should clean up unused statements as it cannot depend upon the close occurring.

Reply:
But there's possibly also scenario where the server doesn't hold resources for a prepared statement (for example, if the handle were self contained), and it seems to me it's a case where the client SHOULD close the prepared statement but where the server MAY close it for external reason (low on resources, timeout, etc...)